### PR TITLE
Console color fun project

### PR DESCRIPTION
**Pull request by cbielow**
## Description

A 'small' fun project, which allows coloring console output.
Colors supported are:
```
/// Text colors/styles supported by Colorizer
    RED,
    GREEN,
    YELLOW,
    BLUE,
    MAGENTA,
    CYAN,
    UNDERLINE,
    BRIGHT, ///< keeps the foreground color, but makes it brighter
    INVERT, ///< invert foreground and background color (inverting twice stays inverted)
```

The output for `usage` of TOPP/UTILS is colored now, as well as all OPENMS_LOG_XXXX streams.
LOG_ERROR is colored red,
LOG_WARN is colored yellow,
LOG_DEBUG is colored magenta

The Colorizer (who is responsible for coloring) will detect if the target stream is std::cout or std::cerr and if the respective stream is actually visible on the console (or redirected to a file or /dev/null).
Only if the stream is visible, coloring will be applied (otherwise one ends up with ANSI codes in redirected files, which is usually not what you want).

Examples
# before
dc_before

# after
dc_after

The current  TOPP color scheme is only a suggestion. You can play around and be more fancy :)

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Comment by cbielow** [comment]
tests should pass now.
Review anyone? :)

**Comment by timosachsenberg** [comment]
how does the detection work if the stream is redirected? if a third-party tool would call (spawn process) and parse the output of a TOPP tool, is it ensured that it doesn't get the color codes?

File: src/openms/include/OpenMS/CONCEPT/Colorizer.h [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    If std::cout or std::cerr are redirected to a file, then Colorizer will detect this and <emph>not emit</empth>
+    any color/style information. This is to avoid the ANSI color codes showing up in the file instead of being
+    filtered out by the console's color process.
+
+    Note: all OS's we know (Windows, Linux, MacOS) only have a single color configuration for the whole
+          console/terminal, independent of streams. I.e. if you apply a permanent color to std::cout
+          then all subsequent output to std::cerr will also be colored! (unless std::cout is redirected to a file,
+          then coloring std::cout will have no effect, not even when printing to the console using std::cerr).
+    
+    Note: this does not (yet) work for OpenMS LogStreams
```
[the remark this answers is not in the document]
Comment (cbielow) [reply src/openms/include/OpenMS/CONCEPT/Colorizer.h:96]: whops.. outdated.
I'll remove that (the commend bugged me, so I solved it, but forgot to remove the comment :))

**Comment by cbielow** [comment]
yes. No color codes in this case, since `isatty()` will yield `false`

File: src/openms/include/OpenMS/CONCEPT/Colorizer.h (hunk 1 above)
Comment (cbielow) [inline src/openms/include/OpenMS/CONCEPT/Colorizer.h:96]:
```suggestion
```

**Comment by jpfeuffer** [comment]
Bildschirmfoto 2022-07-29 um 17 28 01
macOS. How does it look?

**Comment by cbielow** [comment]
Ok, in general.
Maybe a flush at the end of the cout testing block would clean up a bit, since its very lazy on flushes it seems.

Can you also try Decharger help?

Am 29.07.2022 17:28 schrieb Julianus Pfeuffer ***@***.***>:

[Bildschirmfoto 2022-07-29 um 17 28 01]<https://user-images.githubusercontent.com/8102638/181792970-f4b7390c-9bd0-437d-a039-c47670e6c159.png>
macOS. How does it look?

**Comment by cbielow** [comment]
just do clarify: screenshots for MacOSX were from within an IDE -- which apparently has some bugs with buffering streams.
Nothing we can do, so merge.

**Pull request merged**